Lib/test/test_freeze.py

```diff
@@ -0,0 +1,473 @@
+        self.obj.g[1].h = True
+        self.obj.g["two"].i = False
+        freeze(self.obj)
+
+    def test_immutable(self):
+        self.assertTrue(isimmutable(self.obj))
+        self.assertTrue(isimmutable(self.obj.a))
+        self.assertTrue(isimmutable(self.obj.a.b))
+        self.assertTrue(isimmutable(self.obj.d))
+        self.assertTrue(isimmutable(self.obj.d[0]))
+        self.assertTrue(isimmutable(self.obj.d[0].e))
+        self.assertTrue(isimmutable(self.obj.g))
+        self.assertTrue(isimmutable(self.obj.g[1]))
+        self.assertTrue(isimmutable(self.obj.g[1].h))
+        self.assertTrue(isimmutable(self.obj.g["two"]))
+        self.assertTrue(isimmutable(self.obj.g["two"].i))
+
+    def test_set_const(self):
+        with self.assertRaises(NotWriteableError):
+            self.obj.const = 1
+
+    def test_type_immutable(self):
+        self.assertTrue(isimmutable(type(self.obj)))
+        self.assertTrue(isimmutable(type(self.obj).const))
+
+
+class TestFunctions(unittest.TestCase):
+    def setUp(self):
+        def a():
+            return 1
+
+        self.obj = a
+        freeze(self.obj)
+
+    def test_new_function(self):
+        def b():
+            return 1
+
+        self.assertEqual(b(), 1)
+
+    def test_nonlocal(self):
+        def c():
+            v = 0
+
+            def inc():
+                nonlocal v
+                v += 1
+                return v
+
+            return inc
+
+        test = c()
+        self.assertEqual(test(), 1)
+        self.assertEqual(test(), 2)
+        freeze(test)
+        self.assertRaises(NotWriteableError, test)
+
+    def test_global(self):
+        def d():
+            global global0
+            global0 += 1
+            return global0
+
+        self.assertEqual(d(), 1)
+        freeze(d)
+        self.assertTrue(isimmutable(global0))
+        self.assertFalse(isimmutable(global_canary))
+        self.assertRaises(NotWriteableError, d)
+
+    def test_builtins(self):
+        def e():
+            test = list(range(5))
+            return sum(test)
+
+        freeze(e)
+        self.assertTrue(isimmutable(list))
+        self.assertTrue(isimmutable(range))
+        self.assertTrue(isimmutable(sum))
+
+    def test_builtins_nested(self):
+        def g():
+            def nested_test():
+                test = list(range(10))
+                return sum(test)
+
+            return nested_test()
+
+        freeze(g)
+        self.assertTrue(isimmutable(list))
+        self.assertTrue(isimmutable(range))
+        self.assertTrue(isimmutable(sum))
+
+    def test_global_fun(self):
+        def d():
+            return global1_inc()
+
+        freeze(d)
+        self.assertTrue(isimmutable(global1))
+        self.assertTrue(isimmutable(global1_inc))
+        self.assertFalse(isimmutable(global_canary))
+        self.assertRaises(NotWriteableError, d)
+
+    def test_globals_copy(self):
+        def f():
+            global global0
+            return global0
```

Comment:
The special handling for globals uses `PyDict_SetItem`, this means that multiple references to the same value will overwrite the previous value of `frozen_globals`. This should be totally fine, since the old and new value should be identical.

Could modify this test to reference `global0` multiple times, something like this:
```suggestion
    def test_globals_copy(self):
        def f():
            global global0
            ref_1 = global0
            ref_2 = global0
            return global0
```

I mainly want to ensure that the RC and so on remains correct after this :)

The same could be done for builtins, but if this works the others should as well.